### ISSUE TITLE
eQIP Webservices Design Ideas

### ISSUE THREAD

**Issue by floresj** [truetandem/e-QIP-prototype#2196]
This issue contains implementation ideas for integrating with the eQIP web service. We received the  *e-QIP Web Services Developer Manual* which contains the following artifacts:

- `eQIPwebservices.jar` file that can be used as a client within a java application
- API specification documentation. This includes `WSDL` and `javadoc` for the jar client
- Documentation for connecting without the java client
- Schemas for different lookup values 

Our backend services are built on golang which means that _direct_ access to the `eQIPwebservices.jar` is not possible. However, there are some approaches we can take that range from an entirely custom golang SOAP solution to a hybrid approach where we can utilize the `eQIPwebservices.jar`.

The following are some possible approaches:

## 1) Full Golang Solution
This approach requires us to manually create `go` data structures and services that will directly communicate with the eQIP web service. It will be our responsibility to adhere to schema, create the appropriate security tokens using the generated certificates, handle requests/responses and handle errors.

**Advantages**
Full native go solution that requires no integration with external applications. This gives us the most control with handling requests and responses. Since we have the `eQIPwebservices.jar`, I have been able to run some tests to simulate the execution of that jar and see the exact xml that is sent across. This makes it easier for us to define our data structures that map to xml entities/attributes. Plus, how cool would it be to say we have a `go` eQIP ws client.

**Disadvantages**
SOAP communication between different clients on different platforms (from experience) is inherently painful. There are code generation tools that assist with the process but they're usually never complete solutions. However, if we're supporting only a small subset of the operations on the `wsdl`, I don't foresee this being _too_ painful.

## 2) Create wrapper client that Golang executes as an external command
This approach requires a small terminal java app to be created (which I'm currently doing for testing purposes) that can be executed from golang. The `eQIPwebservices.jar` is not directly executable so it is meant to be imported into a java application. This wrapper client would include `eQIPwebservices.jar` 
 and would directly use the classes provided to facilitate the web service communication. Think of this wrapper client as a conduit between golang and the functions in the `eQIPwebservices.jar`. 

Example usage

```golang
   // Execute wrapper client from backend
   cmd := exec.Command("java", "-jar", pathToWrapperClient, put, args, here)
    
   // Java wrapper could write json to the standard output that is then marshaled into something useful
   output, _ := cmd.Output()

   // Take standard output and store into struct
   var resp WebserviceResp
   json.Unmarshal(output,  &resp)
```
**Advantages**:
We indirectly utilize the `eQIPwebservices.jar` so we're not concerned with properly generating XML elements and manually handling requests/responses. We can say with a high degree of certainty that the functions in the `eQIPwebservices.jar` will just work. We simply just need to feed in the required information and let the client handle the details. In this case, we would need to ensure the wrapper client writes any response/error information to the standard output in a format that golang can consume (maybe like json).

**Disadvantages**
The more moving parts the more complicated things get. This would require maintaining a separate wrapper java application. Also, executing external commands from our backend code may not always be a good thing.

### 2.A Microservice-ish (Alternative to terminal client)
This is a long shot but just a thought. Rather than having the wrapper client executed as an external command, it could be run as a java app with an embedded webserver. So golang could communicate with it via a rest api rather than directly executing as a command.

**Comment by ctro** [comment]
Wow, thanks for this great writeup @floresj.  I almost always aim for the simplest thing.  Here, I'm not sure if the simplest thing is writing code for one SOAP call in `go`, or wrapping the `jar` and just using it.

I think my vote is to wrap the `jar` -- this seems like the most straightforward.  It _would_ be cool to have a `go` client :) and, maybe we go there once we're sure that this whole eQIP SOAP API thing is really going to pan out after all?

For this, I'm also v. happy to go with what @floresj and @bmallred feel in their gut.
Also tagging @anthonygarvan for his input

**Comment by bmallred** [comment]
There is no `java` on the cloud.gov buildpack we use. (confirmed with `apt list --installed`)

For the short term I would go with whichever is feasible in the time frame we have.

For the long term I would not go with Java. Besides my distaste for it maintaining security updates for Java in general is a mess in my experience.

**Comment by ctro** [comment]
> For the short term I would go with whichever is feasible in the time frame we have.
> For the long term I would not go with Java

:+1:

**Comment by floresj** [comment]
**Update on Webservices**

I went the route of working with an all `golang` solution. I used the provided jar to determine exact xml structure to best come up with the request and response structure on the `golang` side. Currently, this client has  `isAlive` and  `importRequest` methods. 

In terms of creating the request xml structure, I used the `eQIPwebservices.jar` to send requests to a local golang server that would then just dump the response body. There, I was able to see the structure for the different calls. This made creating the golang structs very easy to model that schema. 

Since we don't have an actual eqip server to simulate what the response structure would look like, I had to read the fun wsdl. However, to further validate, I had that golang webserver respond to the requests made by the `eQIPwebservices.jar` so see if they would be processed appropriately by the jar. With this approach, I can then more accurately model what the golang structs expect to receive.

I also accounted for some error handling. However, the documentation says

> Important: This is still being developed and may not document all possible error messages.

It does provide some error messages that may be returned for the various webservice calls but it is not exhaustive. And it doesn't exactly specify where in the fault element these messages will appear. But this is something that we can get more clarification as we proceed.
